packages/flutter_tools/lib/src/ios/xcode_debug.dart

```diff
@@ -0,0 +1,498 @@
+  final FileSystem _fileSystem;
+  final UserMessages _userMessage;
+  final String _flutterRoot;
+
+  /// Process to start Xcode's debug action.
+  @visibleForTesting
+  Process? startDebugActionProcess;
+
+  /// Information about the project that is currently being debugged.
+  @visibleForTesting
+  XcodeDebugProject? currentDebuggingProject;
+
+  /// Whether the debug action has been started.
+  bool get debugStarted => currentDebuggingProject != null;
+
+  String get pathToXcodeApp {
+    // If the Xcode Select Path is /Applications/Xcode.app/Contents/Developer,
+    // the path to Xcode App is /Applications/Xcode.app
+
+    final String? pathToXcode = _xcode.xcodeSelectPath;
+    if (pathToXcode == null || pathToXcode.isEmpty) {
+      throwToolExit(_userMessage.xcodeMissing);
+    }
+    final int index = pathToXcode.indexOf('.app');
+    if (index == -1) {
+      throwToolExit(_userMessage.xcodeMissing);
+    }
+    return pathToXcode.substring(0, index + 4);
+  }
+
+  String get pathToXcodeAutomationScript {
+    final String flutterToolsAbsolutePath = _fileSystem.path.join(
+      _flutterRoot,
+      'packages',
+      'flutter_tools',
+    );
+    return '$flutterToolsAbsolutePath/bin/xcode_debug.js';
+  }
+
+  /// Install, launch, and start a debug session for app through Xcode interface,
+  /// automated by OSA scripting. First checks if the project is opened in
+  /// Xcode. If it isn't, open it with the `open` command.
+  ///
+  /// The OSA script waits until the project is opened and the debug action
+  /// has started. It does not wait for the app to install, launch, or start
+  /// the debug session.
+  Future<bool> debugApp({
+    required XcodeDebugProject project,
+    required String deviceId,
+    required List<String> launchArguments,
+  }) async {
+
+    // If project is not already opened in Xcode, open it.
+    if (!await _isProjectOpenInXcode(project: project)) {
+      final bool openResult = await _openProjectInXcode(xcodeWorkspace: project.xcodeWorkspace);
+      if (!openResult) {
+        return openResult;
+      }
+    }
+
+    currentDebuggingProject = project;
+    StreamSubscription<String>? stdoutSubscription;
+    StreamSubscription<String>? stderrSubscription;
+    try {
+      startDebugActionProcess = await _processUtils.start(
+        <String>[
+          ..._xcode.xcrunCommand(),
+          'osascript',
+          '-l',
+          'JavaScript',
+          pathToXcodeAutomationScript,
+          'debug',
+          '--xcode-path',
+          pathToXcodeApp,
+          '--project-path',
+          project.xcodeProject.path,
+          '--workspace-path',
+          project.xcodeWorkspace.path,
+          '--device-id',
+          deviceId,
+          '--scheme',
+          project.scheme,
+          '--skip-building',
+          '--launch-args',
+          json.encode(launchArguments),
+          if (project.verboseLogging) '--verbose',
+        ],
+      );
+
+      String stdout = '';
```

Comment:
nit: use StringBuffer:

```suggestion
      StringBuffer stdoutBuffer = StringBuffer();
```